ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java

```diff
@@ -310,20 +310,23 @@ CompletableFuture<Long> getAppliedIndexFuture() {
       return appliedIndexFuture;
     }
 
-    boolean isApplied(LogEntryProto logEntry) {
+    boolean isFirstLogInCurrentTerm(LogEntryProto logEntry) {
       if (appliedIndexFuture.isDone()) {
-        return true;
+        return false;
       }
-      final long appliedIndex = logEntry != null? logEntry.getIndex(): server.getState().getLastAppliedIndex();
-      if (appliedIndex >= startIndex) {
+      final long appliedIndex = logEntry.getIndex();
+      if (appliedIndex == startIndex) {
         appliedIndexFuture.complete(appliedIndex);
-        LOG.info("leader is ready since appliedIndex == {} >= startIndex == {}",
-            appliedIndex, startIndex);
+        LOG.info("leader is ready since appliedIndex == startIndex == {}", startIndex);
         return true;
       } else {
         return false;
       }
     }
```

Comment:
Let's call  it `checkStartIndex`.  We may simplify this method as below:
```java
    boolean checkStartIndex(LogEntryProto logEntry) {
      final boolean completed = logEntry.getIndex() == startIndex && appliedIndexFuture.complete(startIndex);
      if (completed) {
        LOG.info("Leader {} is ready since appliedIndex == startIndex == {}", LeaderStateImpl.this, startIndex);
      }
      return completed;
    }
```

Reply:
Fixed